runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java

```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
+      Map<String, String> inputReplacements = new HashMap<>();
+
+      // Will contain the outputs of the upgraded transform.
+      Map<TupleTag<?>, PCollection<?>> newOutputs = new HashMap<>();
+
+      if (output instanceof PCollectionTuple) {
+        newOutputs.putAll(((PCollectionTuple) output).getAll());
+        for (Map.Entry<TupleTag<?>, PCollection<?>> entry : newOutputs.entrySet()) {
+          if (entry == null) {
+            throw new IllegalArgumentException(
+                "Found unexpected null entry when iterating the outputs of expanded "
+                    + "ExpandableTransform "
+                    + externalTransform);
+          }
+          if (!appliedPTransform.getOutputs().containsKey(entry.getKey())) {
+            throw new RuntimeException(
+                "Could not find the tag " + entry.getKey() + " in the original set of outputs");
+          }
+          PCollection<?> originalOutputPc = originalOutputs.get(entry.getKey());
+          if (originalOutputPc == null) {
+            throw new IllegalArgumentException(
+                "Original output of transform "
+                    + appliedPTransform
+                    + " with tag "
+                    + entry.getKey()
+                    + " was null");
+          }
+          inputReplacements.put(originalOutputPc.getName(), entry.getValue().getName());
+        }
+      } else if (output instanceof PCollection) {
+        newOutputs.put(new TupleTag<>("temp_main_tag"), (PCollection) output);
+        inputReplacements.put(
+            originalOutputs.get(originalOutputs.keySet().iterator().next()).getName(),
+            ((PCollection) output).getName());
+      } else {
+        throw new RuntimeException("Unexpected output type");
+      }
+
+      // We create a new AppliedPTransform to represent the upgraded transform and register it in an
+      // SdkComponents object.
+      AppliedPTransform<?, ?, ?> updatedAppliedPTransform =
+          AppliedPTransform.of(
+              appliedPTransform.getFullName() + "_external",
+              appliedPTransform.getInputs(),
+              newOutputs,
+              externalTransform,
+              externalTransform.getResourceHints(),
+              appliedPTransform.getPipeline());
+      SdkComponents updatedComponents =
+          SdkComponents.create(
+              runnerAPIpipeline.getComponents(), runnerAPIpipeline.getRequirementsList());
+      String updatedTransformId =
+          updatedComponents.registerPTransform(updatedAppliedPTransform, Collections.emptyList());
+      RunnerApi.Components updatedRunnerApiComponents = updatedComponents.toComponents();
+
+      // Recording input updates to the transforms to refer to the upgraded transform instead of the
+      // old one.
+      // Also recording the newly generated id of the old (overridden) transform in the
+      // updatedRunnerApiComponents.
+      Map<String, Map<String, String>> transformInputUpdates = new HashMap<>();
```

Comment:
Could you document what `transformInputUpdates` contains? String isn't very descriptive.